public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx

```diff
@@ -0,0 +1,217 @@
+import { locationService } from '@grafana/runtime';
+import { appEvents } from 'app/core/core';
+import { contextSrv } from 'app/core/services/context_srv';
+import React, { useEffect, useState } from 'react';
+import { Prompt } from 'react-router-dom';
+import { DashboardModel } from '../../state/DashboardModel';
+import { each, filter, find } from 'lodash';
+import angular from 'angular';
+import { UnsavedChangesModal } from '../SaveDashboard/UnsavedChangesModal';
+import * as H from 'history';
+import { SaveLibraryPanelModal } from 'app/features/library-panels/components/SaveLibraryPanelModal/SaveLibraryPanelModal';
+import { PanelModelWithLibraryPanel } from 'app/features/library-panels/types';
+import { useDispatch } from 'react-redux';
+import { discardPanelChanges } from '../PanelEditor/state/actions';
+import { DashboardSavedEvent } from 'app/types/events';
+
+export interface Props {
+  dashboard: DashboardModel;
+}
+
+interface State {
+  original: object | null;
+  originalPath?: string;
+  modal: PromptModal | null;
+  blockedLocation?: H.Location | null;
+}
+
+enum PromptModal {
+  UnsavedChangesModal,
+  SaveLibraryPanelModal,
+}
+
+export const DashboardPrompt = React.memo(({ dashboard }: Props) => {
+  const [state, setState] = useState<State>({ original: null, modal: null });
+  const dispatch = useDispatch();
+  const { original, originalPath, blockedLocation, modal } = state;
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
```

Comment:
I think a comment for why this needs to be in a timeout would be useful

Reply:
yes, (It's to give panels time to load and migrate schemes)